Race when replacing track

<!--
  Come join our community on Slack https://pion.ly/slack for quicker support and more resources

  We also provide a Knowledge Base https://pion.ly/knowledge-base that covers Pion and general WebRTC
  
  We maintain a FAQ of common issues and questions https://pion.ly/knowledge-base/pion-basics/faq/

 Please answer these questions before submitting your issue. Thanks!
-->

### Your environment.
  - Version: master (7b1a63243a47e08f713360a481b86861e2d4a650)
  - Browser: no browser

### What did you do?
<!--
If possible, provide an example for reproducing the error.
A complete runnable program is good.
-->
When replacing tracks multiple threads, AddTrack can return the same free transceiver and try to establish different tracks in it.

Test
```go
func TestRaceReplaceTrack(t *testing.T) {
	lim := test.TimeOut(time.Second * 30)
	defer lim.Stop()

	pc, err := NewPeerConnection(Configuration{})
	assert.NoError(t, err)

	addTrack := func() {
		track, err := NewTrackLocalStaticSample(RTPCodecCapability{MimeType: "video/vp8"}, "foo", "bar")
		assert.NoError(t, err)
		_, err = pc.AddTrack(track) // line 710 invalid state change in RTPTransceiver.setSending
		assert.NoError(t, err)
	}

	for i := 0; i < 100; i++ {
		addTrack()
	}
	for _, tr := range pc.GetTransceivers() {
		assert.NoError(t, pc.RemoveTrack(tr.Sender()))
	}

	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			addTrack() // line 724
			wg.Done()
		}()
	}

	wg.Wait()
	assert.Equal(t, 100, len(pc.GetTransceivers()))
}
```

### What did you expect?
```log
Running tool: C:\Go\bin\go.exe test -timeout 120s -run ^TestRaceReplaceTrack$ github.com/pion/webrtc/v3

--- FAIL: TestRaceReplaceTrack (0.08s)
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:710: 
        	Error Trace:	peerconnection_media_test.go:710
        	            				peerconnection_media_test.go:724
        	            				asm_amd64.s:1373
        	Error:      	Received unexpected error:
        	            	invalid state change in RTPTransceiver.setSending
        	Test:       	TestRaceReplaceTrack
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:710: 
        	Error Trace:	peerconnection_media_test.go:710
        	            				peerconnection_media_test.go:724
        	            				asm_amd64.s:1373
        	Error:      	Received unexpected error:
        	            	invalid state change in RTPTransceiver.setSending
        	Test:       	TestRaceReplaceTrack
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:710: 
        	Error Trace:	peerconnection_media_test.go:710
        	            				peerconnection_media_test.go:724
        	            				asm_amd64.s:1373
        	Error:      	Received unexpected error:
        	            	invalid state change in RTPTransceiver.setSending
        	Test:       	TestRaceReplaceTrack
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:710: 
        	Error Trace:	peerconnection_media_test.go:710
        	            				peerconnection_media_test.go:724
        	            				asm_amd64.s:1373
        	Error:      	Received unexpected error:
        	            	invalid state change in RTPTransceiver.setSending
        	Test:       	TestRaceReplaceTrack

FAIL
FAIL	github.com/pion/webrtc/v3	0.374s
FAIL
```

### What happened?
Replacing all tracks without errors

I have verified that indeed the previous test does not throw an error, but it still exists. Changed the test to check if all the tracks were added at the end. 

Tested on the latest master branch (https://github.com/pion/webrtc/commit/fb988c578132508782313494429bc6a6243726d3), the bug still remains.

```go
func TestRaceReplaceTrack(t *testing.T) {
	pc, err := NewPeerConnection(Configuration{})
	assert.NoError(t, err)

	addTrack := func() *TrackLocalStaticSample {
		track, err := NewTrackLocalStaticSample(RTPCodecCapability{MimeType: "video/vp8"}, "foo", "bar")
		assert.NoError(t, err)
		_, err = pc.AddTrack(track)
		assert.NoError(t, err)
		return track
	}

	for i := 0; i < 10; i++ {
		addTrack()
	}
	for _, tr := range pc.GetTransceivers() {
		assert.NoError(t, pc.RemoveTrack(tr.Sender()))
	}

	var wg sync.WaitGroup
	tracks := make([]*TrackLocalStaticSample, 10)
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func(j int) {
			tracks[j] = addTrack()
			wg.Done()
		}(i)
	}

	wg.Wait()

	for _, track := range tracks {
		have := false
		for _, t := range pc.GetTransceivers() {
			if t.Sender() != nil && t.Sender().Track() == track {
				have = true
				break
			}
		}
		if !have {
			t.Errorf("track was added but not found on senders")
		}
	}

	assert.NoError(t, pc.Close())
}
```

```log
Running tool: C:\Users\dell\sdk\go1.15.10\bin\go.exe test -timeout 120s -run ^TestRaceReplaceTrack$ github.com/pion/webrtc/v3

--- FAIL: TestRaceReplaceTrack (0.06s)
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
    d:\projects\github.com\pion\webrtc\peerconnection_media_test.go:1125: track was added but not found on senders
FAIL
FAIL	github.com/pion/webrtc/v3	0.454s
FAIL
```

--- a/README.md
+++ b/README.md
@@ -266,6 +266,7 @@ Check out the **[contributing wiki](https://github.com/pion/webrtc/wiki/Contribu
 * [Reese](https://github.com/figadore)
 * [David Zhao](https://github.com/davidzhao)
 * [Nam V. Do](https://github.com/namvdo)
+* [Markus Tzoe](https://github.com/zyxar)
 
 ### License
 MIT License - see [LICENSE](LICENSE) for full text

--- a/peerconnection.go
+++ b/peerconnection.go
@@ -571,7 +571,10 @@ func (pc *PeerConnection) getStatsID() string {
 }
 
 func (pc *PeerConnection) hasLocalDescriptionChanged(desc *SessionDescription) bool {
-	for _, t := range pc.GetTransceivers() {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	for _, t := range pc.rtpTransceivers {
 		m := getByMid(t.Mid(), desc)
 		if m == nil {
 			return true
@@ -1532,11 +1535,10 @@ func (pc *PeerConnection) ICEConnectionState() ICEConnectionState {
 }
 
 // GetSenders returns the RTPSender that are currently attached to this PeerConnection
-func (pc *PeerConnection) GetSenders() []*RTPSender {
+func (pc *PeerConnection) GetSenders() (result []*RTPSender) {
 	pc.mu.Lock()
 	defer pc.mu.Unlock()
 
-	result := []*RTPSender{}
 	for _, transceiver := range pc.rtpTransceivers {
 		if transceiver.Sender() != nil {
 			result = append(result, transceiver.Sender())
@@ -1573,7 +1575,8 @@ func (pc *PeerConnection) AddTrack(track TrackLocal) (*RTPSender, error) {
 	}
 
 	var transceiver *RTPTransceiver
-	for _, t := range pc.GetTransceivers() {
+	pc.mu.Lock()
+	for _, t := range pc.rtpTransceivers {
 		if !t.stopped && t.kind == track.Kind() && t.Sender() == nil {
 			transceiver = t
 			break
@@ -1581,19 +1584,23 @@ func (pc *PeerConnection) AddTrack(track TrackLocal) (*RTPSender, error) {
 	}
 	if transceiver != nil {
 		sender, err := pc.api.NewRTPSender(track, pc.dtlsTransport)
+		if err == nil {
+			err = transceiver.SetSender(sender, track)
+			if err != nil {
+				_ = sender.Stop()
+				transceiver.setSender(nil)
+			}
+		}
+		pc.mu.Unlock()
 		if err != nil {
 			return nil, err
 		}
-		transceiver.setSender(sender)
-		// we still need to call setSendingTrack to ensure direction has changed
-		if err := transceiver.setSendingTrack(track); err != nil {
-			return nil, err
-		}
-		pc.onNegotiationNeeded()
 
+		pc.onNegotiationNeeded()
 		return sender, nil
 	}
 
+	pc.mu.Unlock()
 	transceiver, err := pc.AddTransceiverFromTrack(track)
 	if err != nil {
 		return nil, err
@@ -1603,31 +1610,30 @@ func (pc *PeerConnection) AddTrack(track TrackLocal) (*RTPSender, error) {
 }
 
 // RemoveTrack removes a Track from the PeerConnection
-func (pc *PeerConnection) RemoveTrack(sender *RTPSender) error {
+func (pc *PeerConnection) RemoveTrack(sender *RTPSender) (err error) {
 	if pc.isClosed.get() {
 		return &rtcerr.InvalidStateError{Err: ErrConnectionClosed}
 	}
 
 	var transceiver *RTPTransceiver
-	for _, t := range pc.GetTransceivers() {
+	pc.mu.Lock()
+	for _, t := range pc.rtpTransceivers {
 		if t.Sender() == sender {
 			transceiver = t
 			break
 		}
 	}
-
 	if transceiver == nil {
-		return &rtcerr.InvalidAccessError{Err: ErrSenderNotCreatedByConnection}
-	} else if err := sender.Stop(); err != nil {
-		return err
+		err = &rtcerr.InvalidAccessError{Err: ErrSenderNotCreatedByConnection}
+	} else if err = sender.Stop(); err == nil {
+		err = transceiver.setSendingTrack(nil)
 	}
-
-	if err := transceiver.setSendingTrack(nil); err != nil {
+	pc.mu.Unlock()
+	if err != nil {
 		return err
 	}
 
 	pc.onNegotiationNeeded()
-
 	return nil
 }
 
@@ -1855,11 +1861,13 @@ func (pc *PeerConnection) Close() error {
 	closeErrs = append(closeErrs, pc.api.interceptor.Close())
 
 	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #4)
-	for _, t := range pc.GetTransceivers() {
+	pc.mu.Lock()
+	for _, t := range pc.rtpTransceivers {
 		if !t.stopped {
 			closeErrs = append(closeErrs, t.Stop())
 		}
 	}
+	pc.mu.Unlock()
 
 	// https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-close (step #5)
 	pc.sctpTransport.lock.Lock()

--- a/peerconnection_media_test.go
+++ b/peerconnection_media_test.go
@@ -1081,3 +1081,51 @@ func TestPeerConnection_CreateOffer_InfiniteLoop(t *testing.T) {
 
 	assert.NoError(t, pc.Close())
 }
+
+// Assert that AddTrack is thread-safe
+func TestPeerConnection_RaceReplaceTrack(t *testing.T) {
+	pc, err := NewPeerConnection(Configuration{})
+	assert.NoError(t, err)
+
+	addTrack := func() *TrackLocalStaticSample {
+		track, err := NewTrackLocalStaticSample(RTPCodecCapability{MimeType: "video/vp8"}, "foo", "bar")
+		assert.NoError(t, err)
+		_, err = pc.AddTrack(track)
+		assert.NoError(t, err)
+		return track
+	}
+
+	for i := 0; i < 10; i++ {
+		addTrack()
+	}
+	for _, tr := range pc.GetTransceivers() {
+		assert.NoError(t, pc.RemoveTrack(tr.Sender()))
+	}
+
+	var wg sync.WaitGroup
+	tracks := make([]*TrackLocalStaticSample, 10)
+	wg.Add(10)
+	for i := 0; i < 10; i++ {
+		go func(j int) {
+			tracks[j] = addTrack()
+			wg.Done()
+		}(i)
+	}
+
+	wg.Wait()
+
+	for _, track := range tracks {
+		have := false
+		for _, t := range pc.GetTransceivers() {
+			if t.Sender() != nil && t.Sender().Track() == track {
+				have = true
+				break
+			}
+		}
+		if !have {
+			t.Errorf("track was added but not found on senders")
+		}
+	}
+
+	assert.NoError(t, pc.Close())
+}

--- a/sdp.go
+++ b/sdp.go
@@ -400,13 +400,13 @@ func populateSDP(d *sdp.SessionDescription, isPlanB bool, dtlsFingerprints []DTL
 		}
 
 		shouldAddID := true
-		shouldAddCanidates := i == 0
+		shouldAddCandidates := i == 0
 		if m.data {
-			if err = addDataMediaSection(d, shouldAddCanidates, mediaDtlsFingerprints, m.id, iceParams, candidates, connectionRole, iceGatheringState); err != nil {
+			if err = addDataMediaSection(d, shouldAddCandidates, mediaDtlsFingerprints, m.id, iceParams, candidates, connectionRole, iceGatheringState); err != nil {
 				return nil, err
 			}
 		} else {
-			shouldAddID, err = addTransceiverSDP(d, isPlanB, shouldAddCanidates, mediaDtlsFingerprints, mediaEngine, m.id, iceParams, candidates, connectionRole, iceGatheringState, m)
+			shouldAddID, err = addTransceiverSDP(d, isPlanB, shouldAddCandidates, mediaDtlsFingerprints, mediaEngine, m.id, iceParams, candidates, connectionRole, iceGatheringState, m)
 			if err != nil {
 				return nil, err
 			}

